packages/ddp-server/livedata_server.js

```diff
@@ -1643,25 +1653,34 @@ _.extend(Server.prototype, {
       throw new Error("Can't call setUserId on a server initiated method call");
     };
     var connection = null;
-    var currentInvocation = DDP._CurrentInvocation.get();
-    if (currentInvocation) {
-      userId = currentInvocation.userId;
+    var currentMethodInvocation = DDP._CurrentMethodInvocation.get();
+    var currentPublicationInvocation = DDP._CurrentPublicationInvocation.get();
+    var randomSeed = null;
+    if (currentMethodInvocation) {
+      userId = currentMethodInvocation.userId;
+      setUserId = function(userId) {
+        currentMethodInvocation.setUserId(userId);
+      };
+      connection = currentMethodInvocation.connection;
+      randomSeed = DDPCommon.makeRpcSeed(currentMethodInvocation, name);
+    } else if (currentPublicationInvocation) {
+      userId = currentPublicationInvocation.userId;
       setUserId = function(userId) {
-        currentInvocation.setUserId(userId);
+        currentPublicationInvocation._session._setUserId(userId);
```

Comment:
From what I can tell a `Subscription` always have `_session` which is a `Session` and `_setUserId` on `Session` is what's used by a `methodInvocation` to set it's connections `userId`. We need this to be called properly so that publications are re-run when changing the `userId`.

I haven't found a different way of doing this, however maybe we could cc someone at MDG that knows more about the DDP protocol implementation to verify that `_session` is always available before merging?